docs/proposals/20250513-propogate-taints.md

```diff
@@ -0,0 +1,467 @@
+  - Example: Nodes where only GPU related workload should run
+  - CAPI controllers should ensure via reconciliation that such a taint gets added again in case a third party removes it.
+- Taint's supposed to get added once to a Node aka. `Initialize`
+  - This taints are supposed to be set **once** on the `Node` object.
+  - Example: Ensure that no workload gets scheduled to a `Node` unless the taint got removed to e.g. install a GPU driver before allowing workload.
+  - CAPI should once set the taint on the Node and not add it again if it got removed.
+
+#### Proposed API changes
+
+##### Type definition of a taint in Cluster API objects
+
+Note: To reduce verbosity, this proposal does not include all kinds of validation markers.
+
+The following defines a new struct which should be used to define taints at the corresponding API types.
+It replicates the upstream corev1.Taint specification and extends it by a field called `propagation`, which will define the propagation mechanism to use for the taint.
+Using a type definition allows to be extensible and add additional propagation mechanisms when necessary.
+
+```golang
+// MachineTaint defines a taint equivalent to corev1.Taint, but additionally having a propagation field.
+type MachineTaint struct {
+  // key is the taint key to be applied to a Node.
+  // +required
+  Key string `json:"key"`
+
+  // value is the taint value corresponding to the taint key.
+  // +optional
+  Value string `json:"value,omitempty"`
+
+  // effect is the effect for the taint. Valid values are NoSchedule, PreferNoSchedule and NoExecute.
+  // +required
+  // +kubebuilder:validation:Enum=NoSchedule;PreferNoSchedule;NoExecute
+  Effect corev1.TaintEffect `json:"effect"`
+
+  // propagation defines how this taint should be propagated to Nodes.
+  // Always: The taint will be continuously reconciled. If removed from a node, it will be re-added.
+  // OnInitialization: The taint will be added during node initialization. If removed, it will not be re-added.
+  // +required
+  Propagation MachineTaintPropagation `json:"propagation"`
+}
+
+// MachineTaintPropagation defines when a taint should be propagated to Nodes.
+// +kubebuilder:validation:Enum=Always;Initialize
+type MachineTaintPropagation string
+
+const (
+  // TaintPropagationAlways means the taint should be continuously reconciled and kept on the Node.
+  // If removed, it will be re-added during reconciliation.
+  TaintPropagationAlways MachineTaintPropagation = "Always"
+
+  // TaintPropagationOnInitialization means the taint should be set once during initialization and then
+  // left alone. If removed, it will not be re-added.
+  TaintPropagationOnInitialization MachineTaintPropagation = "OnInitialization"
+)
+```
+
+Proper validations on the new field should ensure that no taint with a key of `node.cluster.x-k8s.io/uninitialized` or `node.cluster.x-k8s.io/outdated-revision` is getting added, because that one is restricted to be used by CAPI and providers.
```

Comment:
q: should we reserve the node.cluster.x-k8s.io "namespace" for CAPI, so we can add more CAPI taints in future without worrying of conflict with users managed taints?

Reply:
That was this one:

> I was thinking about that and then I realized that for labels/annotations per default we ~ only propagate "node.cluster.x-k8s.io" (+ some other well-known domains) which would make this a weird inconsistency for taints :/

https://github.com/kubernetes-sigs/cluster-api/pull/12329#discussion_r2438819722